File: Assignment05/List.fs
(* CHANGED FOR ASSIGNMENT 5 *)
// Exercise 5.1
// A) Implement an F# function merge : int list * int list -> int list
// Two sorted lists of integers should be merged into one sorted list of integers

let rec merge (xs, ys) =
    match xs with
    | [] -> ys
    | x :: xs' ->
        match ys with
        | [] -> xs
        | y :: ys' when x <= y -> x :: (merge (xs', ys))
        | y :: ys' -> y :: (merge (xs, ys'));;


merge ([3;5;12], [2;3;4;7]);;
// should give [2;3;3;4;5;7;12]

File: Assignment05/List.cs
// (* CHANGED FOR ASSIGNMENT 5 *)
// Exercise 5.1
// B) Implement a similar method in C# as to Exercise 5.1 A) from List.fs
using System;

public class Program
{
    public static void Main()
    {
        int[] xs = { 3, 5, 12 };
        int[] ys = { 2, 3, 4, 7 };

        var mergelist = merge(xs, ys);
		foreach (var x in mergelist)
		{
			Console.WriteLine(x);
		}
			
    }

    public static int[] merge(int[] xs, int[] ys)
    {
        int xi = 0, yi = 0;
        int[] newlist = new int[xs.Length + ys.Length];
        while (xi + yi < newlist.Length)
        {
            if (xi < xs.Length)
            {
                if (yi < ys.Length && ys[yi] < xs[xi])
                {
                    newlist[yi+xi] = ys[yi++];
                }
                else
                {
                    newlist[yi+xi] = xs[xi++];
                }
            }
            else
            {
                newlist[xi+yi] = ys[yi++];
            }
        }
		return newlist;
    }
}